Upgrade hoedown to 3.0.4

Eljay opened the pull request:
Some hoedown FFI changes:
- `HOEDOWN_EXT_NO_INTRA_EMPHASIS` constant changed.
- Updated/tidied up all callback function signatures.
- All opaque data access has an additional layer of indirection for some reason (`hoedown_renderer_data`).

This also fixes #27862.

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

tamird commented on src/librustdoc/html/markdown.rs:
```diff
@@ -68,49 +67,60 @@ const HOEDOWN_EXTENSIONS: libc::c_uint =
 type hoedown_document = libc::c_void;  // this is opaque to us
```
while you're here, this should probably be an empty enum

daboross replied:
Isn't c_void basically an empty enum?

tamird replied:
i believe that `c_void` is inhabitable; that is, you can create one

alexcrichton replied:
The actual representation or whether this can be constructed isn't so much the problem here, but rather `*mut hoedown_document` and `*mut c_void` are the same types. In Rust it's more idiomatic to have those be separate types (to get at least *some* level of type safety). Creating an empty enum (`enum hoedown_document {}`) should serve this purpose

tamird commented on src/librustdoc/html/markdown.rs:
```diff
@@ -68,49 +67,60 @@ const HOEDOWN_EXTENSIONS: libc::c_uint =
+
+#[repr(C)]
+struct hoedown_renderer_data {
+    opaque: *mut libc::c_void,
```
`opaque` should probably also be an empty enum, but that might be more work

rkruppe replied:
It's a `void *` in the C code as well, so I think this is fine (or even more appropriate, since pointers to empty enums aren't represented as `*i8` in LLVM, while `*mut c_void` is).

tamird commented on src/librustdoc/html/markdown.rs:
```diff
@@ -68,49 +67,60 @@ const HOEDOWN_EXTENSIONS: libc::c_uint =
 
 #[repr(C)]
 struct hoedown_renderer {
     opaque: *mut libc::c_void,
```
ditto

Eljay commented:
Changed `hoedown_document` to be an empty enum and squashed commits.

Adding any kind of type safety to the other void pointers is non-trivial because they point to different types in different passes. A lot of this module is very unsafe/hacky but hopefully we can replace it all with a pure rust Markdown parser at some point anyway!

alexcrichton commented:
@bors: r+ efc98fab26b54ce24e329dff85ca4cbfe3ad8bea

Thanks!

bors commented:
:hourglass: Testing commit efc98fa with merge fd7344c...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/6194), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/6209), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/6149), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6174), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/6152), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/6226), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/6221), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/6185), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/1168), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/1167), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/1166), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/1170), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/450), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/1148)
